src/mesh.cpp

```diff
@@ -3308,10 +3329,26 @@ void LibMesh::initialize()
   if (adaptive_) {
     bin_to_elem_map_.reserve(m_->n_active_elem());
     elem_to_bin_map_.resize(m_->n_elem(), -1);
+
+    //reseve the hash map for cluster elements
+    clustering_element_mapping_.reserve(m_->n_active_elem());
+
+    //adding clustering map
     for (auto it = m_->active_elements_begin(); it != m_->active_elements_end();
-         it++) {
+      it++) {
       auto elem = *it;
 
+      if (amalgamation_) {
+        auto cluster_elem = elem;
+        unsigned int  cluster_id = elem->get_extra_integer(cluster_element_integer_index_);
+        if (cluster_id != -1) {
+          auto first_element_in_a_cluster = m_->elem_ptr(cluster_id);
+          if (first_element_in_a_cluster and first_element_in_a_cluster->active())
+            cluster_elem = first_element_in_a_cluster;
+        }
+        clustering_element_mapping_.insert(std::make_pair(elem, cluster_elem));
+      }
```

Comment:
Put the amalgamation related stuff in the same new API call to initialize amalgamation